ddl/ddl_tiflash_test.go

```diff
@@ -764,3 +767,177 @@ func TestAlterDatabaseBasic(t *testing.T) {
+	// The following statement shall fail, because it reaches limit
+	timeOut, err := execWithTimeout(t, tk, time.Second*3, fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	require.NoError(t, err)
+	require.True(t, timeOut)
+
+	// There must be one table with no TiFlashReplica.
+	time.Sleep(ddl.PollTiFlashInterval * RoundToBeAvailable * 2)
+	check := func(expected int, total int) {
+		cnt := 0
+		for i := 0; i < total; i++ {
+			tb, err := s.dom.InfoSchema().TableByName(model.NewCIStr("tiflash_ddl_limit"), model.NewCIStr(fmt.Sprintf("t%v", i)))
+			require.NoError(t, err)
+			if tb.Meta().TiFlashReplica != nil {
+				cnt += 1
+			}
+		}
+		require.Equal(t, expected, cnt)
+	}
+	check(2, 3)
+	// Wait until timed context quit
+	time.Sleep(time.Second * 2)
+
+	// If we exec in another session, it will not trigger limit. Since DefTiDBBatchPendingTiFlashCount is more than 3.
+	tk2 := testkit.NewTestKit(t, s.store)
+	tk2.MustExec(fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	time.Sleep(ddl.PollTiFlashInterval * RoundToBeAvailable * 1)
+	check(3, 3)
+	time.Sleep(time.Second * 2)
+
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/FastFailCheckTiFlashPendingTables", `return(2)`))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/FastFailCheckTiFlashPendingTables"))
+	}()
+	// This DDL can finish in 3 seconds, since we will force trigger its DDL to update schema cache.
+	tk.MustExec(fmt.Sprintf("create table tiflash_ddl_limit.t%v(z int)", threshold+1))
+	timeOut, err = execWithTimeout(t, tk, time.Second*4, fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	require.NoError(t, err)
+	require.False(t, timeOut)
+	check(4, 4)
+	time.Sleep(time.Second * 2)
+
+	// However, we still have force check next time.
+	tk.MustExec(fmt.Sprintf("create table tiflash_ddl_limit.t%v(z int)", threshold+2))
+	timeOut, err = execWithTimeout(t, tk, time.Millisecond*500, fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	require.NoError(t, err)
+	require.True(t, timeOut)
+	check(4, 5)
+	time.Sleep(time.Second * 2)
+
+	// Retrigger, but close session before the whole job ends.
+	go func() {
+		time.Sleep(time.Millisecond * 20)
+		tk.Session().Close()
+		logutil.BgLogger().Info("session closed")
+	}()
+	timeOut, err = execWithTimeout(t, tk, time.Second*4, fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	require.NoError(t, err)
+	require.False(t, timeOut)
+	check(5, 5)
+	time.Sleep(time.Second * 2)
+}
+
+func TestTiFlashBatchKill(t *testing.T) {
+	s, teardown := createTiFlashContext(t)
+	defer teardown()
+	tk := testkit.NewTestKit(t, s.store)
+
+	threshold := 0
+	tk.MustExec("create database tiflash_ddl_limit")
+	tk.MustExec(fmt.Sprintf("set SESSION tidb_batch_pending_tiflash_count=%v", threshold))
+	tk.MustExec(fmt.Sprintf("create table tiflash_ddl_limit.t%v(z int)", threshold+1))
+
+	go func() {
```

Comment:
Please use `sync.WaitGroup` to make sure we do this logic. And using `WaitGroupWrapper` will be simpler.